staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+	// Interface if possible. It is "possible" if: (1) There's no local interface associated
+	// with the attachment and (2) either there's no remote interface associated with the
+	// attachment, or there's one with the same vni as the one of the attachment.
+	added := ca.addRemoteAttToVNState(nsn.Name, vni)
+	if added {
+		return ca.attemptToCreateRemoteAttIfc(att)
+	}
+
+	return nil
+}
+
+func (ca *ConnectionAgent) attemptToCreateLocalAttIfc(att *netv1a1.NetworkAttachment) (ifcForAttStatus string,
+	ifcForAttStatusValid bool,
+	err error) {
+
+	vni := att.Spec.VNI
+	hostIP := ca.hostIP
+	guestIP := gonet.ParseIP(att.Status.IPv4)
+	guestMAC := generateMACAddr(vni, guestIP)
+	newIfc := netfabric.NetworkInterface{
+		Name:     generateIfcName(guestMAC),
+		VNI:      vni,
+		GuestMAC: guestMAC,
+		GuestIP:  guestIP,
+		HostIP:   hostIP,
+	}
+
+	attNSN := attNSN(att)
+	oldIfc, oldIfcExists, newIfcSet := ca.testAndSwapLocalAttIfcIfVNIsMatch(attNSN, newIfc)
+	if !newIfcSet {
+		err = fmt.Errorf("Attachment %s: found interface with vni %d, attempting to create local one with vni %d. Dropping attempt",
+			attNSN,
+			oldIfc.VNI,
+			vni,
+		)
+		return
+	}
+
+	_, attHasRemoteIfc := ca.getRemoteAttIfc(attNSN)
+	if attHasRemoteIfc {
+		if oldIfcExists {
+			ca.testAndSwapLocalAttIfcIfVNIsMatch(attNSN, oldIfc)
+		} else {
+			ca.removeLocalAttIfc(attNSN)
+		}
+		err = fmt.Errorf("Attachment %s: Attempting to create local interface, found a remote one. Dropping attempt", attNSN)
+		return
+	}
+
+	newIfcNeedsToBeCreated := !oldIfcExists || ifcsDiffer(oldIfc, newIfc)
+	if !newIfcNeedsToBeCreated {
+		// If we're here the old ifc exists and it's equal to the new one
+		ifcForAttStatus = oldIfc.Name
+		ifcForAttStatusValid = true
+		return
+	}
+
+	if oldIfcExists {
+		// If we're here there's an old interface which differs from the new one.
+		// We delete the old interface before creating a new one.
+		err = ca.netFabric.DeleteLocalIfc(oldIfc)
+		if err != nil {
+			ca.testAndSwapLocalAttIfcIfVNIsMatch(attNSN, oldIfc)
+			return
+		}
+	}
+	err = ca.netFabric.CreateLocalIfc(newIfc)
+	if err != nil {
+		ca.removeLocalAttIfc(attNSN)
+		return
+	}
+	ifcForAttStatus = newIfc.Name
+	ifcForAttStatusValid = true
+	return
+}
+
+func (ca *ConnectionAgent) attemptToCreateRemoteAttIfc(att *netv1a1.NetworkAttachment) error {
+	vni := att.Spec.VNI
+	hostIP := gonet.ParseIP(att.Status.HostIP)
+	guestIP := gonet.ParseIP(att.Status.IPv4)
+	guestMAC := generateMACAddr(vni, guestIP)
+	newIfc := netfabric.NetworkInterface{
+		Name:     generateIfcName(guestMAC),
+		VNI:      vni,
+		GuestMAC: guestMAC,
+		GuestIP:  guestIP,
+		HostIP:   hostIP,
+	}
+
+	attNSN := attNSN(att)
+	oldIfc, oldIfcExists, newIfcSet := ca.testAndSwapRemoteAttIfcIfVNIsMatch(attNSN, newIfc)
+	if !newIfcSet {
+		return fmt.Errorf("Attachment %s: found interface with vni %d, attempting to create remote one with vni %d. Dropping attempt",
```

Comment:
Also, this is not necessarily going to react well when notifcation and handling happens in an unexpected order.  Consider: create with VNI1, delete, create with VNI2.  Supposed both creations get worked on concurrently.  Suppose creation on VNI1 gets in and sets the VNI.  Then creation of VNI2 is processed; it will be ignored here.  Later deletion at VNI1 will be processed.  But nothing revisits the creation on VNI2.

Reply:
The fact that VNI1 passes through and sets the VNI causes testAndSwapRemoteAttIfcIfVNIsMatch to return the flag newIfcSet with a false value when acting on VNI2. This causes an error to be returned, which in turn causes the reference with VNI2 to be enqueued again, hence it is not ignored.